File: 2020/day8/day8.py
#!/usr/bin/env python3
# https://adventofcode.com/2020/day/8

import sys
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
lines = data.splitlines()


def run(program):
    acc = 0
    ip = 0
    visited = set()
    while ip < len(program) and ip not in visited:
        visited.add(ip)

        instr = program[ip]
        ip += 1

        op, imm = instr.split(maxsplit=1)
        if op == "nop":
            pass
        elif op == "acc":
            acc += int(imm)
        elif op == "jmp":
            ip += int(imm) - 1
        else:
            raise ValueError(instr)

    return acc, ip == len(program)


# part 1
acc, booted = run(lines)
assert booted is False
print(acc)


# part 2
for i in range(len(lines)):
    backup = lines[i]

    instr, imm = backup.split(maxsplit=1)
    if instr == "nop":
        lines[i] = f"jmp {imm}"
    elif instr == "jmp":
        lines[i] = f"nop {imm}"

    acc, booted = run(lines)
    if booted:
        break

    lines[i] = backup

print(acc)

File: 2020/day8/day8.rs
//! [Day 8: Handheld Halting](https://adventofcode.com/2020/day/8)

use std::collections::HashSet;

fn run(boot_code: &[String]) -> (i32, bool) {
    let mut acc = 0;
    let mut ip = 0;
    let mut seen = HashSet::new();

    while ip < boot_code.len() && seen.insert(ip) {
        let instr = &boot_code[ip];
        ip += 1;

        let (op, imm) = instr.split_once(' ').unwrap();

        match op {
            "nop" => continue,
            "acc" => {
                acc += imm.parse::<i32>().unwrap();
            }
            "jmp" => {
                ip = ip.wrapping_add_signed(imm.parse::<isize>().unwrap()) - 1;
            }
            _ => panic!("invalid instr {instr}"),
        };
    }

    (acc, ip == boot_code.len())
}

struct Puzzle {
    boot_code: Vec<String>,
}

impl Puzzle {
    const fn new() -> Self {
        Self {
            boot_code: Vec::new(),
        }
    }

    /// Get the puzzle input.
    fn configure(&mut self, path: &str) {
        let data = std::fs::read_to_string(path).unwrap_or_else(|_| {
            eprintln!("cannot read input file {path}");
            std::process::exit(1);
        });

        self.boot_code = data.lines().map(std::string::ToString::to_string).collect();
    }

    /// Solve part one.
    fn part1(&self) -> i32 {
        run(&self.boot_code).0
    }

    /// Solve part two.
    fn part2(&self) -> i32 {
        let mut boot_code = self.boot_code.clone();
        for ip in 0..boot_code.len() {
            let instr = &boot_code[ip].to_string();

            let (op, imm) = instr.split_once(' ').unwrap();

            if op == "acc" {
                continue;
            }

            boot_code[ip] = if op == "nop" {
                format!("jmp {imm}")
            } else {
                format!("nop {imm}")
            };

            let (acc, terminated) = run(&boot_code);
            if terminated {
                return acc;
            }

            boot_code[ip] = instr.to_string();
        }
        0
    }
}

fn main() {
    let args = aoc::parse_args();
    let mut puzzle = Puzzle::new();
    puzzle.configure(args.path.as_str());
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_part1() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test.txt");
        assert_eq!(puzzle.part1(), 5);
    }

    #[test]
    fn test_part2() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test.txt");
        assert_eq!(puzzle.part2(), 8);
    }
}
